File: LeetCode/138CopyListWithRandomPointer/CopyListWithRandomPointer.cpp
class Solution {
public:
    Node* copyRandomList(Node* head) {
        // 2 passes -> Time:O(n), Space:O(n)
        map<Node*,Node*> m;
        m[NULL] = NULL;
        Node *copy, *curr = head;
        while (curr) {
            m[curr] = new Node(curr->val);
            curr = curr->next;
        }
        curr = head;
        while(curr) {
            copy = m[curr];
            copy->next = m[curr->next];
            copy->random = m[curr->random];
            curr = curr->next;
        }
        return m[head];
    }
};

File: LeetCode/138CopyListWithRandomPointer/CopyListWithRandomPointer.py
class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        # 2 passes -> Time:O(n), Space:O(n)
        original_copy = {None:None}
        curr = head
        while curr:
            copy = Node(curr.val)
            original_copy[curr] = copy
            curr = curr.next
        curr = head
        while curr:
            copy = original_copy[curr]
            copy.next = original_copy[curr.next]
            copy.random = original_copy[curr.random]
            curr = curr.next
        return original_copy[head]
